### PR TITLE
Atomically release quota and queue inadmissible workloads

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Ensures atomicity when releasing quota and requeuing inadmissible workloads. If not, it's possible that a workload being requeued slips in between the time a workload is deleted from the cache and we requeue inadmissible workloads, violating priority.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

This is adding a dependency from the queue manager lock into the cache lock. We already had an instance like this via the StatusChecker interface in the manager.
In the future, we need to be careful not to add a lock dependency the other way around.

Based on top of #506

### CODE REVIEW

File: pkg/controller/core/workload_controller.go [hunk 1]
```diff
@@ -154,14 +154,15 @@ func (r *WorkloadReconciler) Delete(e event.DeleteEvent) bool {
...
-		if err := r.cache.DeleteWorkload(wl); err != nil {
-			if !e.DeleteStateUnknown {
-				log.Error(err, "Failed to delete workload from cache")
-			}
-		}
 
 		// trigger the move of associated inadmissibleWorkloads if required.
-		r.queues.QueueAssociatedInadmissibleWorkloads(ctx, wl)
+		r.queues.QueueAssociatedInadmissibleWorkloadsAfter(ctx, wl, func() {
+			if err := r.cache.DeleteWorkload(wl); err != nil {
```
Comment (ahg-g) [inline pkg/controller/core/workload_controller.go:160]: You may want to add a comment here on why we are doing this: The workload is deleted from the cache while holding the queue lock to ensure that potentially high priority workloads are requeued first, and then the deleted workload's resources are released. This prevents a lower priority workload waiting in the queue from obtaining the released resources before the high priority ones.

File: pkg/queue/manager.go [hunk 2]
```diff
@@ -341,19 +341,20 @@ func (m *Manager) deleteWorkloadFromQueueAndClusterQueue(w *kueue.Workload, qKey
...
 }
 
-// QueueAssociatedInadmissibleWorkloads moves all associated workloads from
-// inadmissibleWorkloads to heap. If at least one workload is moved,
-// returns true. Otherwise returns false.
-func (m *Manager) QueueAssociatedInadmissibleWorkloads(ctx context.Context, w *kueue.Workload) {
+// QueueAssociatedInadmissibleWorkloadsAfter moves all associated workloads from
+// inadmissibleWorkloads to the heaps, after executing action, while holding the
+// lock. If at least one workload is moved, returns true. Otherwise returns false.
+func (m *Manager) QueueAssociatedInadmissibleWorkloadsAfter(ctx context.Context, w *kueue.Workload, action func()) {
```
Comment (ahg-g) [inline pkg/queue/manager.go:347]: You changed the semantics of this function. Add a comment that requeueing happens only if the workload triggering it is an admitted one?

Comment (alculquicondor) [reply pkg/queue/manager.go:350]: In theory this is not changing the behavior, as we would only call this function for admitted workloads. So it's more of a code simplification.
But good thing to note in the documentation.

**Pull request merged**